cupy/_core/core.pyx

```diff
@@ -2579,22 +2644,91 @@ cdef _ndarray_base _array_from_nested_cupy_sequence(
     return a
 
 
+cdef inline _ndarray_base _try_skip_h2d_copy(
+        obj, dtype, bint copy, order, Py_ssize_t ndmin):
+    if copy:
+        return None
+
+    if not is_sam_supported(device.get_device_id()):
+        return None
+
+    if not isinstance(obj, numpy.ndarray):
+        return None
+
+    # dtype should not change
+    obj_dtype = obj.dtype
+    if not (obj_dtype == get_dtype(dtype) if dtype is not None else True):
+        return None
+
+    # CuPy onlt supports numerical dtypes
+    if obj_dtype.char not in _dtype.all_type_chars:
+        return None
+
+    # CUDA onlt supports little endianness
+    if obj_dtype.byteorder not in ('|', '=', '<'):
+        return None
+
+    # strides and the requested order could mismatch
+    obj_flags = obj.flags
+    if not internal._is_layout_expected(
+            obj_flags.c_contiguous, obj_flags.f_contiguous, order):
+        return None
+
+    cdef intptr_t ptr = obj.ctypes.data
+#    # NumPy ndarrays do not guarantee alignment unless a custom
+#    # allocator is in use
+#    if not _is_alignment_expected(ptr, obj_dtype.itemsize):
+#        return None
+
+    # NumPy 0-size arrays still have non-null pointers...
+    cdef size_t nbytes = obj.nbytes
+    if nbytes == 0:
+        ptr = 0
+
+    cdef Py_ssize_t ndim = obj.ndim
+    cdef tuple shape = obj.shape
+    cdef tuple strides = obj.strides
+    if ndmin > ndim:
+        # pad shape & strides
+        shape = (1,) * (ndmin - ndim) + shape
+        strides = (shape[0] * strides[0],) * (ndmin - ndim) + strides
+
+    cdef memory.SystemMemory ext_mem = memory.SystemMemory.from_external(
+        ptr, nbytes, obj)
+    cdef memory.MemoryPointer memptr = memory.MemoryPointer(ext_mem, 0)
+    return ndarray(shape, obj_dtype, memptr, strides)
+
+
 cdef _ndarray_base _array_default(
-        obj, dtype, order, Py_ssize_t ndmin, bint blocking):
+        obj, dtype, bint copy, order, Py_ssize_t ndmin, bint blocking):
+    cdef _ndarray_base a
+
+    # Fast path: zero-copy a NumPy array if possible
+    if not blocking:
+        a = _try_skip_h2d_copy(obj, dtype, copy, order, ndmin)
```

Comment:
Note: we were wondering if this can be called unconditionally, regardless of `blocking`. Should be revisited in another PR.